### PR TITLE
transformers: knowledge distillation mixin and initial experiments

### PR DESCRIPTION
**Pull request by lucasosouza**
Initial distillation mixin to transformers. Results with tiny bert are encouraging (<-- means "distilled from"): 

| model           | num params | train loss  | eval loss      | perplexity |
|-----------------|:----------:|:-----------:|:--------------:|:----------:|
| mini_bert_100k  | 10,615,808 | 3.669       | 4.773          | 27.342     |
| mini_bert_50k   | 10,615,808 | 4.464       | 5.815          | 56.309     |
| tiny_bert_50k <-- bert_1mi   |  4,124,928 | -       | 5.351          | 40.817    |
| tiny_bert_100k  |  4,124,928 | 4.566       | 5.802          | 55.793   |
| tiny_bert_50k   |  4,124,928 | 5.990       | 8.367          | 330.234    |

Learning rate in the distillation experiment is two orders of magnitude higher, and we can probably increase it even more.

Missing a few things, main are: 
* unit tests
* combining target with teacher softmax outputs (there is a small bug to fix there, likely related to the extra context dimension)
* a lot more experimentation with higher LR schedules and ensembles

### CODE REVIEW

**Review by mvacaporale — APPROVED** [review]
Looks good. Only minor/nit-pick comments.

File: projects/transformers/trainer_mixins/distillation.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        kd_temperature = self.compute_kd_temperature(trainer_state)
+
+        def get_logits(output):
+            return output["logits"] if isinstance(output, dict) else output[0]
+
+        student_logits = get_logits(model_output)
+
+        with torch.no_grad():
+            # If ensemble, linearly combine outputs of softmax
+            softmax_output_teacher = None
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/trainer_mixins/distillation.py:294]: I did have a `torch.zeros_like` (keep in mind this is a tensor) and then the update. But it is faster like this. If I am using a single teacher, I was able to improve it/sec by 5-10% by using this approach.

A lot of decisions made in this KD code in general have to do with faster computation and less memory, evaluated through empirical evaluation, although I am sure there are more places we can refactor and squeeze to get better performance. Changing every possible operation to be inplace is one case (for example `t.abs_()` is an order of magnitude faster than `t = t.abs()`)

File: projects/transformers/trainer_mixins/distillation.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+        return setattr(rgetattr(obj, pre) if pre else obj, post, val)
+
+    def rgetattr(obj, attr, *args):
+        def _getattr(obj, attr):
+            return getattr(obj, attr, *args)
+        return functools.reduce(_getattr, [obj] + attr.split("."))
+
+    # Find and replace all position embeddings if the size of position embedding
+    # doesn't match new sequence length
+    for module_name, module in model.named_modules():
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/trainer_mixins/distillation.py:413]: That is good to know, thanks. They seem to do the same thing. You did showed me this function before, but I didn't make the connection while coding that this would be the perfect use case.

File: projects/transformers/trainer_mixins/distillation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/trainer_mixins/distillation.py:294]: I see. That's interesting. I wouldn't have thought such small things could make a difference in computation time (like not using `torch.zeros_like` when it's not needed). But it makes sense as well.

File: projects/transformers/trainer_mixins/distillation.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+def resize_position_embeddings(model, new_seq_length):
+    """
+    Resizes model's position embeddings matrices if the size of max position embedding
+    doesn't match new sequence length.
+    (size of position embedding equals size of the attention window)
+
+    :param new_seq_length: Tokenizer sequence length.
+    """
+
+    position_embeddings = filter_modules(
```
Comment (lucasosouza) [inline projects/transformers/trainer_mixins/distillation.py:404]: @mvacaporale changed to use the existing functions. Tested on a notebook, seems to work as before, haven't tested beyond that.

Comment (lucasosouza) [inline projects/transformers/trainer_mixins/distillation.py:404]: @mvacaporale changed to use the existing functions. Tested on a notebook, seems to work as before, haven't tested beyond that.
